File: backend/app/api/prescription.py
"""
Prescription API - NLP-based prescription generation and translation
"""

from fastapi import APIRouter, HTTPException, Depends
from sqlalchemy.orm import Session
from sqlalchemy import select
from pydantic import BaseModel
from typing import Optional, List, Dict
from datetime import datetime
import uuid

from app.ml.nlp_prescriber import NLPPrescriber
from app.ml.translator import BilingualTranslator
from app.core.dependencies import get_current_user
from app.core.database import get_db
from app.models.db_models import User, Diagnosis, Prescription

router = APIRouter()

# Initialize NLP models
_prescriber = None
_translator = None


def get_prescriber():
    global _prescriber
    if _prescriber is None:
        _prescriber = NLPPrescriber()
    return _prescriber


def get_translator():
    global _translator
    if _translator is None:
        _translator = BilingualTranslator()
    return _translator


class Medication(BaseModel):
    name: str
    type: str
    dosage: str
    frequency: str
    duration: str
    instructions: str
    warnings: Optional[List[str]] = None


class PrescriptionRequest(BaseModel):
    diagnosis_id: str
    additional_notes: Optional[str] = None


class PrescriptionResponse(BaseModel):
    id: str
    diagnosis_id: str
    severity: str
    medications: List[Medication]
    lifestyle_recommendations: List[str]
    follow_up_instructions: str
    reasoning: str
    status: str
    created_at: str


class TranslateRequest(BaseModel):
    prescription_id: str
    target_language: str  # "en" or "te"


@router.post("/generate", response_model=PrescriptionResponse)
async def generate_prescription(
    request: PrescriptionRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Generate AI-powered prescription based on diagnosis.
    """
    # Get diagnosis (user-specific)
    diagnosis = db.execute(
        select(Diagnosis).where(
            Diagnosis.id == request.diagnosis_id,
            Diagnosis.user_id == current_user.id
        )
    ).scalar_one_or_none()
    
    if not diagnosis:
        raise HTTPException(status_code=404, detail="Diagnosis not found")
    
    # Check if prescription already exists
    existing = db.execute(
        select(Prescription).where(
            Prescription.diagnosis_id == request.diagnosis_id,
            Prescription.user_id == current_user.id
        )
    ).scalar_one_or_none()
    
    if existing:
        return PrescriptionResponse(
            id=existing.id,
            diagnosis_id=existing.diagnosis_id or "",
            severity=existing.severity,
            medications=existing.medications,
            lifestyle_recommendations=existing.lifestyle_recommendations,
            follow_up_instructions=existing.follow_up_instructions,
            reasoning=existing.reasoning,
            status=existing.status,
            created_at=existing.created_at.isoformat()
        )
    
    # Generate prescription using NLP
    prescriber = get_prescriber()
    prescription_data = prescriber.generate(
        severity=diagnosis.severity,
        lesion_counts=diagnosis.lesion_counts,
        clinical_metadata=diagnosis.clinical_metadata or {},
        additional_notes=request.additional_notes
    )
    
    # Create prescription record
    prescription_id = str(uuid.uuid4())[:8]
    prescription = Prescription(
        id=prescription_id,
        user_id=current_user.id,
        diagnosis_id=request.diagnosis_id,
        severity=diagnosis.severity,
        medications=prescription_data["medications"],
        lifestyle_recommendations=prescription_data["lifestyle_recommendations"],
        follow_up_instructions=prescription_data["follow_up_instructions"],
        reasoning=prescription_data["reasoning"],
        status="generated"
    )
    
    db.add(prescription)
    db.commit()
    db.refresh(prescription)
    
    return PrescriptionResponse(
        id=prescription.id,
        diagnosis_id=prescription.diagnosis_id or "",
        severity=prescription.severity,
        medications=prescription.medications,
        lifestyle_recommendations=prescription.lifestyle_recommendations,
        follow_up_instructions=prescription.follow_up_instructions,
        reasoning=prescription.reasoning,
        status=prescription.status,
        created_at=prescription.created_at.isoformat()
    )


@router.get("/{prescription_id}")
async def get_prescription(
    prescription_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get prescription by ID (user-specific)."""
    prescription = db.execute(
        select(Prescription).where(
            Prescription.id == prescription_id,
            Prescription.user_id == current_user.id
        )
    ).scalar_one_or_none()
    
    if not prescription:
        raise HTTPException(status_code=404, detail="Prescription not found")
    
    return {
        "id": prescription.id,
        "diagnosis_id": prescription.diagnosis_id,
        "severity": prescription.severity,
        "medications": prescription.medications,
        "lifestyle_recommendations": prescription.lifestyle_recommendations,
        "follow_up_instructions": prescription.follow_up_instructions,
        "reasoning": prescription.reasoning,
        "status": prescription.status,
        "created_at": prescription.created_at.isoformat()
    }


@router.get("/")
async def list_prescriptions(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """List all prescriptions for current user."""
    prescriptions = db.execute(
        select(Prescription).where(Prescription.user_id == current_user.id)
        .order_by(Prescription.created_at.desc())
    ).scalars().all()
    
    return [
        {
            "id": p.id,
            "diagnosis_id": p.diagnosis_id,
            "severity": p.severity,
            "medications": p.medications,
            "lifestyle_recommendations": p.lifestyle_recommendations,
            "follow_up_instructions": p.follow_up_instructions,
            "reasoning": p.reasoning,
            "status": p.status,
            "created_at": p.created_at.isoformat()
        }
        for p in prescriptions
    ]


@router.post("/translate")
async def translate_prescription(
    request: TranslateRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Translate prescription to English or Telugu.
    """
    prescription = db.execute(
        select(Prescription).where(
            Prescription.id == request.prescription_id,
            Prescription.user_id == current_user.id
        )
    ).scalar_one_or_none()
    
    if not prescription:
        raise HTTPException(status_code=404, detail="Prescription not found")
    translator = get_translator()
    
    translated = translator.translate_prescription(
        medications=prescription.medications,
        recommendations=prescription.lifestyle_recommendations,
        instructions=prescription.follow_up_instructions,
        target_language=request.target_language
    )
    
    return {
        "prescription_id": request.prescription_id,
        "original_language": "en" if request.target_language == "te" else "te",
        "target_language": request.target_language,
        "translated_content": translated,
        "translated_at": datetime.now().isoformat()
    }

File: backend/src/routes/prescription.ts
import { Router, Request, Response } from 'express';
import { body, validationResult } from 'express-validator';
import { authenticate } from '../middleware/auth';
import { AppDataSource } from '../database/connection';
import { Diagnosis } from '../models/Diagnosis';
import {
  generatePrescription,
  createPrescription,
  getPrescriptionById,
  getPrescriptionByDiagnosisId,
  listPrescriptions,
  translatePrescription,
} from '../services/prescriptionService';

const router = Router();

/**
 * POST /api/v1/prescription/generate
 * Generate AI-powered prescription based on diagnosis
 */
router.post(
  '/generate',
  authenticate,
  [
    body('diagnosis_id').notEmpty().withMessage('diagnosis_id is required'),
    body('additional_notes').optional().isString(),
  ],
  async (req: Request, res: Response) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ detail: errors.array() });
      }

      const { diagnosis_id, additional_notes } = req.body;
      const user = req.user!;

      // Get diagnosis (user-specific)
      const diagnosisRepository = AppDataSource.getRepository(Diagnosis);
      const diagnosis = await diagnosisRepository.findOne({
        where: { id: diagnosis_id, userId: user.id },
      });

      if (!diagnosis) {
        return res.status(404).json({ detail: 'Diagnosis not found' });
      }

      // Check if prescription already exists
      const existing = await getPrescriptionByDiagnosisId(diagnosis_id, user.id);
      if (existing) {
        // Patients can only see approved prescriptions
        if (user.role === 'patient' && existing.status !== 'approved') {
          return res.json({
            id: existing.id,
            status: existing.status,
            message: existing.status === 'pending'
              ? 'Prescription is pending doctor approval'
              : 'Prescription was rejected',
            created_at: existing.createdAt.toISOString(),
          });
        }
        
        return res.json({
          id: existing.id,
          diagnosis_id: existing.diagnosisId || '',
          severity: existing.severity,
          medications: existing.medications,
          lifestyle_recommendations: existing.lifestyleRecommendations,
          follow_up_instructions: existing.followUpInstructions,
          reasoning: existing.reasoning,
          status: existing.status,
          doctor_notes: existing.doctorNotes,
          approved_at: existing.approvedAt?.toISOString(),
          created_at: existing.createdAt.toISOString(),
        });
      }

      // Generate prescription using Gemini API
      const prescriptionData = await generatePrescription(
        diagnosis.severity,
        diagnosis.lesionCounts,
        diagnosis.clinicalMetadata || {},
        diagnosis.problemSummary || undefined,
        additional_notes,
        diagnosis.acneType
      );

      // Create prescription record
      const prescription = await createPrescription(
        user.id,
        diagnosis_id,
        diagnosis.severity,
        prescriptionData
      );

      // Return appropriate response based on user role
      if (user.role === 'patient') {
        // Patients only see status, not full prescription until approved
        res.json({
          id: prescription.id,
          status: prescription.status,
          message: 'Prescription generated and sent for doctor approval. You will be notified once approved.',
          created_at: prescription.createdAt.toISOString(),
        });
      } else {
        // Doctors can see full prescription details
        res.json({
          id: prescription.id,
          diagnosis_id: prescription.diagnosisId || '',
          severity: prescription.severity,
          medications: prescription.medications,
          lifestyle_recommendations: prescription.lifestyleRecommendations,
          follow_up_instructions: prescription.followUpInstructions,
          reasoning: prescription.reasoning,
          status: prescription.status,
          created_at: prescription.createdAt.toISOString(),
        });
      }
    } catch (error) {
      console.error('Generate prescription error:', error);
      res.status(500).json({ detail: 'Internal server error' });
    }
  }
);

/**
 * GET /api/v1/prescription/:id
 * Get prescription by ID
 */
router.get('/:id', authenticate, async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const user = req.user!;

    const prescription = await getPrescriptionById(id, user.id);
    if (!prescription) {
      return res.status(404).json({ detail: 'Prescription not found' });
    }

    res.json({
      id: prescription.id,
      diagnosis_id: prescription.diagnosisId,
      severity: prescription.severity,
      medications: prescription.medications,
      lifestyle_recommendations: prescription.lifestyleRecommendations,
      follow_up_instructions: prescription.followUpInstructions,
      reasoning: prescription.reasoning,
      status: prescription.status,
      created_at: prescription.createdAt.toISOString(),
    });
  } catch (error) {
    console.error('Get prescription error:', error);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

/**
 * GET /api/v1/prescription
 * List all prescriptions for current user (all statuses for patients: approved, rejected, pending)
 */
router.get('/', authenticate, async (req: Request, res: Response) => {
  try {
    const user = req.user!;
    const prescriptions = await listPrescriptions(user.id);

    // Patients can see all their prescriptions (approved, rejected, pending)
    // Doctors can see all prescriptions
    // No filtering needed - return all prescriptions for the user

    res.json(
      prescriptions.map((p) => ({
        id: p.id,
        diagnosis_id: p.diagnosisId,
        severity: p.severity,
        medications: p.medications,
        lifestyle_recommendations: p.lifestyleRecommendations,
        follow_up_instructions: p.followUpInstructions,
        reasoning: p.reasoning,
        status: p.status,
        doctor_notes: p.doctorNotes,
        approved_at: p.approvedAt?.toISOString(),
        created_at: p.createdAt.toISOString(),
      }))
    );
  } catch (error) {
    console.error('List prescriptions error:', error);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

/**
 * POST /api/v1/prescription/translate
 * Translate prescription to English or Telugu
 */
router.post(
  '/translate',
  authenticate,
  [
    body('prescription_id').notEmpty(),
    body('target_language').isIn(['en', 'te', 'hi']),
  ],
  async (req: Request, res: Response) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ detail: errors.array() });
      }

      const { prescription_id, target_language } = req.body;
      const user = req.user!;

      const prescription = await getPrescriptionById(prescription_id, user.id);
      if (!prescription) {
        return res.status(404).json({ detail: 'Prescription not found' });
      }

      const translated = await translatePrescription(
        prescription.medications,
        prescription.lifestyleRecommendations,
        prescription.followUpInstructions,
        target_language as 'en' | 'te' | 'hi'
      );

      res.json({
        prescription_id,
        original_language: 'en',
        target_language,
        translated_content: {
          medications: translated.medications,
          lifestyle_recommendations: translated.lifestyleRecommendations,
          follow_up_instructions: translated.followUpInstructions,
        },
        translated_at: new Date().toISOString(),
      });
    } catch (error) {
      console.error('Translate prescription error:', error);
      res.status(500).json({ detail: 'Internal server error' });
    }
  }
);

export default router;

